src/core/thread/network_data_leader.cpp

```diff
@@ -577,20 +575,200 @@ void Leader::SendServerDataResponse(const Coap::Header &aRequestHeader, const Ip
     }
 }
 
+void Leader::RlocLookup(uint16_t aRloc16, bool *aIn, bool *aStable, uint8_t *aTlvs, uint8_t aTlvsLength)
+{
+    NetworkDataTlv *cur = reinterpret_cast<NetworkDataTlv *>(aTlvs);
+    NetworkDataTlv *end = reinterpret_cast<NetworkDataTlv *>(aTlvs + aTlvsLength);
+    NetworkDataTlv *subCur;
+    NetworkDataTlv *subEnd;
+    PrefixTlv *prefix;
+    BorderRouterTlv *borderRouter;
+    HasRouteTlv *hasRoute;
+    BorderRouterEntry *borderRouterEntry;
+    HasRouteEntry *hasRouteEntry;
+    *aIn = false;
+    *aStable = false;
+
+    while (1)
+    {
+        if (cur >= end)
+        {
+            break;
+        }
+
+        if (cur->GetType() == NetworkDataTlv::kTypePrefix)
+        {
+            prefix = reinterpret_cast<PrefixTlv *>(cur);
+            subCur = reinterpret_cast<NetworkDataTlv *>(prefix->GetSubTlvs());
+            subEnd = reinterpret_cast<NetworkDataTlv *>(prefix->GetSubTlvs() + prefix->GetSubTlvsLength());
+
+            while (1)
+            {
+                if (subCur >= subEnd)
```

Comment:
Why not put `subCur < subEnd` in the `while` statement above and remove this `if` statement?